File: bench_uds.py
import socket
import time
import os
import multiprocessing

SOCKET_PATH = "/tmp/unix_socket_benchmark"
PACKET_SIZE = 1 * 3 * 16 * 112 * 112 * 4  # Compute the packet size
NUM_ITERATIONS = 5  # Number of messages to send

def server():
    if os.path.exists(SOCKET_PATH):
        os.remove(SOCKET_PATH)
    
    server_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    server_socket.bind(SOCKET_PATH)
    server_socket.listen(1)
    
    conn, _ = server_socket.accept()
    while True:
        buf = []
        while len(buf) < PACKET_SIZE:
            data = conn.recv(PACKET_SIZE - len(buf))
            buf.extend(data)
        print("recd with len ", len(buf))
        if not buf:
            break
        conn.sendall(bytes(buf))  # Echo back
    
    conn.close()
    server_socket.close()
    os.remove(SOCKET_PATH)

def client():
    time.sleep(1)  # Give server some time to start
    client_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    client_socket.connect(SOCKET_PATH)
    
    data = b"0" * PACKET_SIZE  # Create a dummy payload
    
    # Benchmark latency
    start_time = time.time()
    for i in range(NUM_ITERATIONS):
        client_socket.sendall(data)
        buf = []
        while len(buf) < PACKET_SIZE:
            tmp = client_socket.recv(PACKET_SIZE - len(buf))
            buf.extend(tmp)
        print(f"Done {i+1}")
    end_time = time.time()
    
    latency = (end_time - start_time) / NUM_ITERATIONS
    throughput = (PACKET_SIZE * NUM_ITERATIONS) / (end_time - start_time) / (1024 * 1024)  # MB/s
    
    print(f"Average Latency: {latency * 1e6:.2f} µs")
    print(f"Throughput: {throughput:.2f} MB/s")
    
    client_socket.close()

def main():
    server_process = multiprocessing.Process(target=server)
    server_process.start()
    
    client()
    
    server_process.terminate()
    server_process.join()

if __name__ == "__main__":
    main()

File: bench_uds.go
package main

import (
	"fmt"
	"net"
	"os"
	"time"
)

const (
	SOCKET_PATH   = "/tmp/unix_socket_benchmark"
	PACKET_SIZE   = 1 * 3 * 16 * 112 * 112 * 4
	NUM_ITERATIONS = 1000
)

func server() {
	os.Remove(SOCKET_PATH)
	ln, err := net.Listen("unix", SOCKET_PATH)
	if err != nil {
		fmt.Println("Error starting server:", err)
		return
	}
	defer ln.Close()
	defer os.Remove(SOCKET_PATH)

	conn, err := ln.Accept()
	if err != nil {
		fmt.Println("Error accepting connection:", err)
		return
	}
	defer conn.Close()

	buffer := make([]byte, PACKET_SIZE)
	for {
		totalRead := 0
		for totalRead < PACKET_SIZE {
			n, err := conn.Read(buffer[totalRead:])
			if err != nil || n == 0 {
				return
			}
			totalRead += n
		}
		conn.Write(buffer[:totalRead]) // Echo back
	}
}

func client() {
	time.Sleep(1 * time.Second) // Give server time to start
	conn, err := net.Dial("unix", SOCKET_PATH)
	if err != nil {
		fmt.Println("Error connecting to server:", err)
		return
	}
	defer conn.Close()

	data := make([]byte, PACKET_SIZE)
	response := make([]byte, PACKET_SIZE)

	start := time.Now()
	for i := 0; i < NUM_ITERATIONS; i++ {
		_, err := conn.Write(data)
		if err != nil {
			fmt.Println("Error sending data:", err)
			return
		}

		totalRead := 0
		for totalRead < PACKET_SIZE {
			n, err := conn.Read(response[totalRead:])
			if err != nil {
				fmt.Println("Error receiving data:", err)
				return
			}
			totalRead += n
		}
	}
	end := time.Now()

	elapsed := end.Sub(start).Seconds()
	latency := (elapsed / float64(NUM_ITERATIONS)) * 1e6 // Convert to microseconds
	throughput := (float64(PACKET_SIZE*NUM_ITERATIONS) / elapsed) / (1024 * 1024) // MB/s

	fmt.Printf("Average Latency: %.2f µs\n", latency)
	fmt.Printf("Throughput: %.2f MB/s\n", throughput)
}

func main() {
	go server()
	time.Sleep(500 * time.Millisecond) // Give server some time to start
	client()
}

